types/wlr_data_control_v1.c

```diff
@@ -5,51 +5,26 @@
 #include <unistd.h>
 #include <wlr/types/wlr_data_control_v1.h>
 #include <wlr/types/wlr_data_device.h>
+#include <wlr/types/wlr_primary_selection.h>
 #include <wlr/util/log.h>
 #include "util/signal.h"
 #include "wlr-data-control-unstable-v1-protocol.h"
 
-#define DATA_CONTROL_MANAGER_VERSION 1
+#define DATA_CONTROL_MANAGER_VERSION 2
 
-struct client_data_source {
-	struct wlr_data_source source;
+struct data_control_source {
 	struct wl_resource *resource;
+	struct wl_array mime_types;
 	bool finalized;
-};
-
-static const struct wlr_data_source_impl client_source_impl;
 
-static struct client_data_source *
-		client_data_source_from_source(struct wlr_data_source *wlr_source) {
-	assert(wlr_source->impl == &client_source_impl);
-	return (struct client_data_source *)wlr_source;
-}
-
-static void client_source_send(struct wlr_data_source *wlr_source,
-		const char *mime_type, int fd) {
-	struct client_data_source *source =
-		client_data_source_from_source(wlr_source);
-	zwlr_data_control_source_v1_send_send(source->resource, mime_type, fd);
-	close(fd);
-}
-
-static void client_source_destroy(struct wlr_data_source *wlr_source) {
-	struct client_data_source *source =
-		client_data_source_from_source(wlr_source);
-	zwlr_data_control_source_v1_send_cancelled(source->resource);
-	// Make the resource inert
-	wl_resource_set_user_data(source->resource, NULL);
-	free(source);
-}
-
-static const struct wlr_data_source_impl client_source_impl = {
-	.send = client_source_send,
-	.destroy = client_source_destroy,
+	// Only one of these is non-NULL.
+	struct wlr_data_source *active_source;
```

Comment:
Do we really need to keep the `data_control_source` around after `set_selection`?

Reply:
Well, it can receive the destroy event, in which case it must clean up the underlying `wlr_data_source` or `wlr_primary_selection_source`.

Reply:
Err, right.